core/consensus/pow/src/lib.rs

```diff
@@ -0,0 +1,399 @@
+};
+use sr_primitives::Justification;
+use sr_primitives::generic::{BlockId, Digest, DigestItem};
+use sr_primitives::traits::{Block as BlockT, Header as HeaderT, ProvideRuntimeApi};
+use srml_timestamp::{TimestampInherentData, InherentError as TIError};
+use pow_primitives::{Difficulty, Seal, POW_ENGINE_ID};
+use primitives::H256;
+use inherents::{InherentDataProviders, InherentData};
+use consensus_common::{
+	BlockImportParams, BlockOrigin, ForkChoiceStrategy,
+	well_known_cache_keys::Id as CacheKeyId, Environment, Proposer,
+};
+use consensus_common::import_queue::{BoxBlockImport, BasicQueue, Verifier};
+use codec::{Encode, Decode};
+use log::*;
+
+/// Auxiliary storage prefix for PoW engine.
+pub const POW_AUX_PREFIX: [u8; 4] = *b"PoW:";
+
+fn aux_key(hash: &H256) -> Vec<u8> {
+	POW_AUX_PREFIX.iter().chain(&hash[..])
+		.cloned().collect::<Vec<_>>()
+}
+
+/// Auxiliary storage data for PoW.
+#[derive(Encode, Decode, Clone, Debug, Default)]
+pub struct PowAux {
+	/// Total difficulty.
+	pub total_difficulty: Difficulty,
+}
+
+impl PowAux {
+	/// Read the auxiliary from client.
+	pub fn read<C: AuxStore>(client: &C, hash: &H256) -> Result<Self, String> {
+		let key = aux_key(hash);
+
+		match client.get_aux(&key).map_err(|e| format!("{:?}", e))? {
+			Some(bytes) => PowAux::decode(&mut &bytes[..]).map_err(|e| format!("{:?}", e)),
+			None => Ok(Default::default()),
+		}
+	}
+}
+
+/// Algorithm used for proof of work.
+pub trait PowAlgorithm<B: BlockT> {
+	/// Get the next block's difficulty.
+	fn difficulty(&self, parent: &BlockId<B>) -> Result<Difficulty, String>;
+	/// Verify proof of work against the given difficulty.
+	fn verify(
+		&self,
+		parent: &BlockId<B>,
+		pre_hash: &H256,
+		seal: &Seal,
+		difficulty: Difficulty,
+	) -> Result<bool, String>;
+	/// Mine a seal that satisfy the given difficulty.
+	fn mine(
+		&self,
+		parent: &BlockId<B>,
+		pre_hash: &H256,
+		seed: &H256,
+		difficulty: Difficulty,
+		round: u32,
+	) -> Result<Option<Seal>, String>;
+}
+
+/// A verifier for PoW blocks.
+pub struct PowVerifier<C, Algorithm> {
+	client: Arc<C>,
+	algorithm: Algorithm,
+	inherent_data_providers: inherents::InherentDataProviders,
+}
+
+impl<C, Algorithm> PowVerifier<C, Algorithm> {
+	fn check_header<B: BlockT<Hash=H256>>(
+		&self,
+		mut header: B::Header,
+		parent_block_id: BlockId<B>,
+	) -> Result<(B::Header, Difficulty, DigestItem<H256>), String> where
+		Algorithm: PowAlgorithm<B>,
+	{
+		let hash = header.hash();
+
+		let (seal, inner_seal) = match header.digest_mut().pop() {
+			Some(DigestItem::Seal(id, seal)) => {
+				if id == POW_ENGINE_ID {
+					(DigestItem::Seal(id, seal.clone()), seal)
+				} else {
+					return Err(format!("Header uses the wrong engine {:?}", id))
+				}
+			},
+			_ => return Err(format!("Header {:?} is unsealed", hash)),
+		};
+
+		let pre_hash = header.hash();
+		let difficulty = self.algorithm.difficulty(&parent_block_id)?;
+
+		if !self.algorithm.verify(
```

Comment:
Where are the signatures checked?  This is not obvious from the code, and it should be.

Reply:
There’s no signature check in PoW engine.